### PR TITLE
DEVAI-140: Initial code drop of CLI code

### PR DESCRIPTION
**Pull request by gabemontero**
Per the team discussions, I've organized the initial drop for this CLI as a single PR, but with different commits for different subsets of this initial drop.  This should hopefully help allow for review of a subset of related files vs. tacking everything at once.
- main-cmd:  the entrypoint for the command
- markdowns-git-make: the initial README and other set of markdown documents, gitignore, makefile
- go-mod-tidy-vendor: the go.mod/go.sum and vendor tree
- followed by breakdowns of package code, unit tests, and your typical util / config / test type of packages

### CODE REVIEW

**Comment by gabemontero** [comment]
I'll add github actions to run the unit tests after this PR merges, and I'll gate future PRs on unit tests needing to pass

File: docs/background.md
Comment (Jdubrick) [inline docs/background.md:38]: Nit: `developer/tester engineer` reads weird to me personally

**Review by Jdubrick — COMMENTED** [review]
Reviewed commit https://github.com/redhat-ai-dev/rhdh-ai-catalog-cli/pull/2/commits/2f5102b054f5e50d2c359cece5e171caa1e1bfcc

File: README.md
Comment (gabemontero) [reply README.md:11]: although we'll be switching to RHDHPAI (I need to make this update) but that will be in the same situation, but yes, that is the initial drawback chosen for the sake of avoiding the need to duplicate, etc.

there will be a socialization effort within the RH ecosystem for the CLI that will start once we have something available, and if it gains traction, I can see us either using a public Jira or turning github issues back on

among other things, per the roadmap.md, here is A LOT of fit and finish that needs to be done regardless ... so some restriction / control on issue creation seem suitable until we can get ahead of that productization.

File: docs/DEVELOPMENT.md
Comment (gabemontero) [reply docs/DEVELOPMENT.md:12]: yep will fix in next update thanks

File: docs/DEVELOPMENT.md
Comment (gabemontero) [reply docs/DEVELOPMENT.md:22]: yep

File: docs/DEVELOPMENT.md
Comment (gabemontero) [reply docs/DEVELOPMENT.md:28]: yep

File: docs/background.md
Comment (gabemontero) [reply docs/background.md:38]: ok I will attempt to reword .... any suggestions on how I should capture both dev and QE ?

File: README.md
Comment (gabemontero) [reply README.md:11]: I'll add a note that once we reach a reasonable level wrt basic fit and finish we will revisit turning on github issues

File: docs/background.md
Comment (Jdubrick) [reply docs/background.md:38]: We could probably swap it to 
```suggestion
As a development/test engineer, I want to administrate the Backstage Catalog from the command line so that I can better automate verification pipelines for testing of Backstage’s AI related features.
```

Or maybe
```suggestion
As a developer/tester, I want to administrate the Backstage Catalog from the command line so that I can better automate verification pipelines for testing of Backstage’s AI related features.
```

Comment (gabemontero) [reply docs/background.md:38]: I think I'm going to go with DevOps

**Comment by gabemontero** [comment]
hey @Jdubrick I've pushed updates from you comments as a separate commit to facilitate review

File: test/stub/rest.go
Comment (Jdubrick) [inline test/stub/rest.go:10]: Am I missing it or are these unused?

**Review by Jdubrick — COMMENTED** [review]
Minor comment for commit https://github.com/redhat-ai-dev/rhdh-ai-catalog-cli/pull/2/commits/13c88b70825a22d19318c074b312cfe47a9f852b

**Review by Jdubrick — COMMENTED** [review]
https://github.com/redhat-ai-dev/rhdh-ai-catalog-cli/pull/2/commits/bf268c2fb88f01388e5e8977144d24f7dba01aa8 and https://github.com/redhat-ai-dev/rhdh-ai-catalog-cli/pull/2/commits/4ed583563ef82e84560ff4cafb96903a5285d0f5 lgtm

Changes made in https://github.com/redhat-ai-dev/rhdh-ai-catalog-cli/pull/2/commits/23e8affe52c75f940c521c68fc598f86441de322 for my comments also lgtm

File: test/stub/rest.go
Comment (gabemontero) [reply test/stub/rest.go:10]: they are used in the unit tests in other packages ..... all three are used 

I just confirmed in my Goland Jetbrains, which in addition to its usage search, has a feature that flags unused const/vars/funcs.

**Review by johnmcollier — COMMENTED** [review]
Honestly, didn't have any big concerns with this, it's pretty much in line with the designs you've shared previously and the walkthrough earlier this week

File: pkg/cmd/cli/backstage/constants.go
Comment (gabemontero) [reply pkg/cmd/cli/backstage/constants.go:6]: So should we bother picking any value for an API's `spec.type` ?  

At least with the current examples on the dev cluster, the `definition` fields start with `openapi` since those AI API happen to be expressed as `openapi`.  So in theory we could in can be inferred from that.

Do you know if the same is true for `asyncapi` or `graphql` or `grpc` @johnmcollier ?

thanks

Comment (johnmcollier) [reply pkg/cmd/cli/backstage/constants.go:6]: I’d assume so.

I’m guessing that we don’t need to store an API type for the CLI, and we just infer about their membership in the model catalog based on what depends on them (e.g. model servers or models)

Comment (gabemontero) [reply pkg/cmd/cli/backstage/constants.go:6]: ok ... it makes the filtering on "APIs that are only related to AI" a bit more interesting 

I'll keep it unset for now, and may just open an additional fit and finish item to sort out filtering for API

File: docs/DEVELOPMENT.md
Comment (gabemontero) [reply docs/DEVELOPMENT.md:5]: changed to bullets ... part of next push

File: pkg/cmd/cli/kserve/kserve.go
Comment (gabemontero) [reply pkg/cmd/cli/kserve/kserve.go:199]: OK I'm just going to reuse the name for now then, and we'll wait for customer feedback to see if some sort of "preferred clean up pattern" emerges.  Given many of these model registries are still evolving, we'll see what transpires on that end as well.

File: pkg/cmd/cli/backstage/kind_api.go
Comment (gabemontero) [reply pkg/cmd/cli/backstage/kind_api.go:8]: The open source project I started with generated all the schema but as we have discussed, those backstream schema files were not kept up to date, and we may want to engage with backstream upstream to find out what the skinny is.  So bottom line we'll need to manually update as we decide to leverage a change made upstream.

for now I'll try and clarify the comment along these lines.

File: pkg/cmd/cli/kserve/kserve.go
Comment (gabemontero) [reply pkg/cmd/cli/kserve/kserve.go:29]:
```
# The '--kubeconfig' flag can be used to set the location of the configuration file used for accessing the credentials
# used for interacting with the Kubernetes cluster.
```
is part of the next push

File: pkg/cmd/cli/backstage/apis_test.go
Comment (gabemontero) [reply pkg/cmd/cli/backstage/apis_test.go:74]: I was getting a little worried about the length of the file.

Let's play this one by ear .... there are plenty of free web pages out there that will create the pretty version for you.

https://jsonformatter.org/json-pretty-print for example handled this JSON just fine

Comment (Jdubrick) [reply pkg/cmd/cli/backstage/apis_test.go:74]: I'm wondering if storing that in a file and just reading it in for the tests would make maintaining it or debugging easier?

File: pkg/cmd/cli/backstage/constants.go
Comment (johnmcollier) [reply pkg/cmd/cli/backstage/constants.go:6]: Yeah, I think that's fine.

Comment (gabemontero) [reply pkg/cmd/cli/backstage/constants.go:6]: OK I found examples of each of them at https://github.com/backstage/backstage/tree/master/plugins/api-docs/dev ... btw it appears to be both `trpc`  and `grpc`

I think then we can just look at the start of the definition string and set this accordingly.  Otherwise we'll leave blank.

And I'll open a new fit and finish search story for fetching apis from ai components and resources.

Part of next push

File: docs/roadmap.md
Comment (amfred) [inline docs/roadmap.md:67]: I'd replace the competitor comment with something like "Most popular source for public models."
It doesn't really matter if they're a competitor.

File: docs/roadmap.md
Comment (amfred) [inline docs/roadmap.md:70]: Can you sort the "high" priority items to the top? So, move OCI up.

File: docs/roadmap.md
Comment (amfred) [inline docs/roadmap.md:8]: Are these roughly in the order you think they should be implemented? That would be nice if they were.

**Review by amfred — COMMENTED** [review]
Several small comments, nothing major. 
Well, renaming a command would be a pain, but you don't have to do it.

File: docs/roadmap.md
Comment (gabemontero) [reply docs/roadmap.md:50]: there is a little bit of common overlap but there is an analogous `glab` tool for Gitlab that basically is their attempt to keep with `gh`

See https://gitlab.com/gitlab-org/cli

Also golang based

You'll see where they basically copied syntax in many cases.

I'll mention that one as well.

File: pkg/cmd/cli/root.go
Comment (gabemontero) [reply pkg/cmd/cli/root.go:19]: A bit, but I went with `new-model` given that OCP users are very used to `oc new-app...` and the like.

That said I can always add it as another alias minimally.  I'll do that since that is trivial.  Long term, let's wait until some external feedback while still in alpha phase for broader name changes.

Comment (gabemontero) [reply pkg/cmd/cli/root.go:19]: But sure I'll add more elaboration / attempt to make clear.

**Comment by gabemontero** [comment]
ok @johnmcollier I've pushed the updates from your commits

File: docs/roadmap.md
Comment (gabemontero) [reply docs/roadmap.md]: No let's just sort out priority when I open the Jiras.  I'm also curious on evengelist feedback to sort out which should be done first.

**Comment by gabemontero** [comment]
ok @amfred I've pushed the updates (as separate commits to allow you to narrow what files you need to look at) for the comments from today - thanks

File: docs/roadmap.md
Comment (amfred) [reply docs/roadmap.md:75]: OK

File: docs/roadmap.md
Comment (amfred) [reply docs/roadmap.md]: OK

**Review by amfred — APPROVED** [review]
LGTM

**Review by johnmcollier — APPROVED** [review]

**Comment by gabemontero** [comment]
OK I've updated the roadmap.md with the initial set of Jira links.

Going ahead and merging ... leaving commits unsquashed to reflect initial review contributions and break up file listings when folks look at git history for a given file in the future.

**Pull request merged**
